### ISSUE TITLE
Windows: cmd.exe + quotes = Hell

### ISSUE THREAD

**Issue by fdcastel** [fsprojects/Fable#27]
In Windows 8.1, with Node.js 5.4.0, running

`node tools\fable2babel.js --projFile "C:\Temp\MyProj\Core\MyProj.Core.fsproj"
`

gives me this

```
cmd /C C:\Temp\Fable\tools/../build/main/Fable.exe {"lib":".","outDir":".","symbols":[],"watch":false,"projFile":"MyProj.Core.fsproj"}
events.js:141
      throw er; // Unhandled 'error' event
      ^

Error: spawn cmd ENOENT
    at exports._errnoException (util.js:856:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:178:32)
    at onErrorNT (internal/child_process.js:344:16)
    at nextTickCallbackWith2Args (node.js:478:9)
    at process._tickCallback (node.js:392:17)
    at Function.Module.runMain (module.js:432:11)
    at startup (node.js:141:18)
    at node.js:1003:3

```

which is not a really useful error message :) -- But then, I tried

`cmd /C C:\Temp\Fable\tools/../build/main/Fable.exe {"lib":".","outDir":".","symbols":[],"watch":false,"projFile":"MyProj.Core.fsproj"}
`

which outputs a somewhat better description of the real problem

```
Unhandled Exception: Newtonsoft.Json.JsonReaderException: Input string '.' is not a valid number. Path 'lib', line 1, position 6.
   at Newtonsoft.Json.JsonTextReader.ParseNumber(ReadType readType)
   at Newtonsoft.Json.JsonTextReader.ReadStringValue(ReadType readType)
   at Newtonsoft.Json.JsonTextReader.ReadAsString()
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ReadForType(JsonReader reader, JsonContract contract, Boolean hasConverter)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ResolvePropertyAndCreatorValues(JsonObjectContract contract, JsonProperty containerProperty, JsonReader reader, Type objectType)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObjectUsingCreatorWithParameters(JsonReader reader, JsonObjectContract contract, JsonProperty containerProperty, ObjectConstructor`1 creator, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateNewObject(JsonReader reader, JsonObjectContract objectContract, JsonProperty containerMember, JsonProperty containerProperty, String id, Boolean& createdFromNonDefaultCreator)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value)
   at Fable.Main.main(String[] argv) in C:\Temp\Fable\src\Main.fs:line 73

```

I can't see how that json argument could work at all in Windows. Those double quotes in json would be parsed by the command line processor (cmd.exe). And trying to escape them would be... troublesome (to say the least -- See http://stackoverflow.com/a/15262019 and http://blogs.msdn.com/b/twistylittlepassagesallalike/archive/2011/04/23/everyone-quotes-arguments-the-wrong-way.aspx )

Or I am missing something?

**Comment by alfonsogarciacaro** [comment]
Thanks for your detailed feedback. I'm not on Windows right now so I didn't notice. The weird thing is that other Windows users didn't report this problem (maybe it has been solved in Windows 10?). I wanted to use JSON because in some situations I pass code directly to `Fabel.exe` and `JSON.stringify` is an easy way to escape problematic characters. Anyway, this will be a rare case so it'll probably be better to pass the arguments in a more classic for general situations.

Thanks again for the report!

**Comment by alfonsogarciacaro** [comment]
I'm passing the arguments in a more sensible way now :) Can you please check if it works now in your computer?

**Comment by rflechner** [comment]
Hi,

I had the same problem on my Windows computer,
so I tested your last commit.

The tool failed due to a invalid path and it couldn't resolve app.fsx ( when I built node server sample )

Here is my workaround to compil it on Windows (if it could be helpful) : https://github.com/rflechner/Fable/commit/f5a7e21d47e13650fa58672e86cdbd6decc20de1

Regards

**Comment by alfonsogarciacaro** [comment]
Nice fix! Can you please send a PR? Thanks!

**Comment by fdcastel** [comment]
Great work guys! After @alfonsogarciacaro and @rflechner fixes I believe it is working (with a minor glitch, see below)

Running

`node tools\fable2babel.js --projFile "C:\Temp\MyProj\Core\MyProj.Core.fsproj"
`

yet gives me the cryptic

```
cmd /C C:\temp\Fable\build\main\Fable.exe --lib . --outDir . --watch false --projFile ./MyProj.Core.fsproj
events.js:154
      throw er; // Unhandled 'error' event
      ^

Error: spawn cmd ENOENT
    at exports._errnoException (util.js:856:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:178:32)
    at onErrorNT (internal/child_process.js:344:16)
    at _combinedTickCallback (node.js:377:13)
    at process._tickCallback (node.js:401:11)
    at Function.Module.runMain (module.js:449:11)
    at startup (node.js:141:18)
    at node.js:933:3

```

But then running

`cmd /C C:\temp\Fable\build\main\Fable.exe --lib . --outDir . --watch false --projFile ./MyProj.Core.fsproj
`

We can see the real problem:

`{"type":"Error","message":"Could not find file 'C:\\temp\\Fable\\MyProj.Core.fsproj'."}
`

fable2babel.js is looking for the .fsproj project in the current folder.

Maybe a pushd / popd before / after the processing? :)


As a workaround, running the script from my project folder 

```
cd \temp\MyProj\Core
node c:\temp\fable\tools\fable2babel.js --projFile .\MyProj.Core.fsproj
```

gives me (which I believe is) the correct output

```
cmd /C c:\temp\fable\build\main\Fable.exe --lib . --outDir . --watch false --projFile ./MyProj.Core.fsproj
Cannot find replacement for Microsoft.FSharp.Collections.Array.fill at C:\temp\MyProj\Core\Memory.fs (18,8--18,55) IsSynthetic=false

```


(Now I just need to find myself around this "Cannot find replacement for Microsoft.FSharp.Collections.Array.fill", but this is an entirely different problem :) )

Again, thanks for the prompt feedback, guys!

**Comment by davidpodhola** [comment]
I am not sure if this is not related to where you run `node` and where is the `fsx` file.
Please have a look at [the smallest console sample](https://github.com/davidpodhola/fable-samples/tree/master/consoleSample) and try to run at your computer with [the path to Fable](https://github.com/davidpodhola/fable-samples/blob/master/consoleSample/build.cmd#L3) changed.
When Fable is packaged, these issues will go away I am sure.

**Comment by fdcastel** [comment]
@davidpodhola: I did run your sample and it worked fine. However, please note that your build.cmd does exactly the same that my last "workaround" does: it calls fable2babel.js from the same folder where your projFile is (in your case, solution.fsx)

The problem is: when I call the script from a different folder

```
cd c:\Temp\Fable
node tools\fable2babel.js --projFile "C:\Temp\MyProj\Core\MyProj.Core.fsproj" 
```

it tries to run this

`cmd /C C:\temp\Fable\build\main\Fable.exe --lib . --outDir . --watch false --projFile ./MyProj.Core.fsproj 
`

loosing the full path for the "MyProj.Core.fsproj" file (it is not in the current dir)

**Comment by davidpodhola** [comment]
Sorry, missed your point about your last workaround.
Please have a look at [different folders sample](https://github.com/davidpodhola/fable-samples/tree/master/differentFolders). Does this work on your environment? Would it fix the issue?

**Comment by fdcastel** [comment]
Yeah! It worked.

```
====================================
Stats

Files:
- Total: 1
- Binary: 0 (skipped)
- With Matches: 0
- Without Matches: 1
- Failed to Open: 0
- Failed to Write: 0

Matches:
- Found: 0
- Replaced: 0

Duration: 0,031 secs
====================================
cmd /C c:\temp\Fable\build\main\Fable.exe --lib c:/temp/Fable/lib --outDir ../out --watch false --projFile ./solution.fsx --env node
Compiled C:\Temp\fable-samples\differentFolders\src\solution.fsx
Finished
```

But I see you are doing some voodoo magic :) with fnr.exe (find-and-replace?) command in your build script and using a fableconfig.json file (which I'm not!)

Please note I'm using the simplest possible thing, as stated in /docs/compiling.md:

> A node script is used to trigger the compilation from F# to JS.
> To execute it, make sure you've installed the dependencies with
> `npm install` and from a terminal on the project root folder, type:
> 
> ```
> node tools/fable2babel.js --projFile paht/to/YourProject.fsproj
> ```

**Comment by davidpodhola** [comment]
Yeah, I just followed the [@alfonsogarciacaro samples](https://github.com/fsprojects/Fable/tree/master/sample/node). Voodoo magic is needed as Fable is not yet released and published to repositories. I assume you will be the able to `npm install fable` later (well, even now, why not?).

**Comment by alfonsogarciacaro** [comment]
Hi there, I've tried on Windows and (hopefully) fixed the problem. No Voodoo magic should be necessary now? Would yo mind trying again (simplest sample, as in docs) and reopen the issue if you're still having problems?

Thanks a lot for your help! And yes, David, as soon as these problems are ironed out I'm planning to distribute the compiler through npm :)

**Comment by fdcastel** [comment]
Thanks @alfonsogarciacaro. Unfortunately, the problem remains :(

I believe it have something to do with the handling of **fableCwd** variable. I will try to describe exactly what I did:

* I did pull the latest sources from your last commit (git pull)
* I run build.cmd to rebuild fable.exe
* my git repo is in C:\temp\fable

Then, from my root git folder (cd c:\temp\fable) i just ran

`node tools\fable2babel.js --projFile "c:\temp\fable\sample\node\server\app.fsx"
`

which outputs

```
cmd /C C:\Temp\Fable\build\main\Fable.exe --lib . --outDir . --watch false --projFile ./app.fsx
events.js:154
      throw er; // Unhandled 'error' event
      ^

Error: spawn cmd ENOENT
    at exports._errnoException (util.js:856:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:178:32)
    at onErrorNT (internal/child_process.js:344:16)
    at _combinedTickCallback (node.js:377:13)
    at process._tickCallback (node.js:401:11)
    at Function.Module.runMain (module.js:449:11)
    at startup (node.js:141:18)
    at node.js:933:3

```

However, just looking to the first line of this output, I'm wondering: Is that "--projFile ./app.fsx" passed to fable.exe right? What is the current directory when this cmd.exe is called?

Looking for this answer, i changed the line 192 of fable2babel.js to 

```
    console.log("fableCwd = " + fableCwd);
    console.log(fableCmd + " " + fableCmdArgs.join(" "));
```

and I ran it again. To my surprise it gave me

```
fableCwd = C:\Temp\Fable/c:\temp\Fable\sample\node\server
cmd /C C:\Temp\Fable\build\main\Fable.exe --lib . --outDir . --watch false --projFile ./app.fsx
events.js:154
      throw er; // Unhandled 'error' event
    (...)

```

Do you see it? The value of fableCwd is completely messed up (it should be just "c:\temp\Fable\sample\node\server" in this case).

Sorry for the crude tips but I'm without a good nodejs debugging environment here at the moment ;)

**Comment by davidpodhola** [comment]
Maybe I am missing something clear, just few remarks:

- I remember having similar problems as @fdcastel maybe until I switched from `\` to `/` as a path separator
- Current version does not work at all for me - my samples break with `The path is not of a legal form`. Tests are running fine. Will add my samples as tests.
- I think until Fable is in `node_modules` or some other well know folder we will have to do "voodoo" to set the right path. I was inspired by how in earlier versions of [Orleans](https://github.com/dotnet/orleans) the compilation was done (before moving everything to nugets).

**Comment by davidpodhola** [comment]
See [PR#32](https://github.com/fsprojects/Fable/pull/32)

**Comment by alfonsogarciacaro** [comment]
Hi @fdcastel, thanks a lot for trying again and for your reports. It seems that node `path.isAbsolute` is not working correctly then (see [here](https://github.com/fsprojects/Fable/blob/master/tools/fable2babel.js#L146)). I tried in Windows 10 and it's working fine, but I've seen in the internet it doesn't with previous versions of Windows (see [here](http://stackoverflow.com/questions/21698906/how-to-check-if-a-path-is-absolute-or-relative)). I need to ask you a last favour, could you please try this in Windows cmd and tell me the results?

```
node
> var p = "c:\temp\fable\sample\node\server\app.fsx" 
> var path = require("path")
> path.isAbsolute(p)
> path.resolve(p) == path.normalize(p)
```
In my machine with Windows 10 both of the last lines return true, but if in your case only the last line returns true (as suggested in the SO response), we could fix the problem just by replacing the method of checking absolute paths.

Cheers!

**Comment by davidpodhola** [comment]
Just for the evidence on my computer (after I changed the path to a `fsx` file), the results are

```
> var p = "C:\Users\david.podhola\S\fable-samples\consoleSample\solution.fsx"
undefined
> var path = require("path")
undefined
> path.isAbsolute(p)
false
> path.resolve(p) == path.normalize(p)
false
>
```
also if I try to change the path I get

```
> var q = "/Users/david.podhola/S/fable-samples/consoleSample/solution.fsx"
undefined
> path.isAbsolute(q)
true
>  path.resolve(q) == path.normalize(q)
false

```

**Comment by alfonsogarciacaro** [comment]
Damn! Why I'm getting different results in my computer then? Maybe because I'm using Windows on Bootcamp (Mac)? Anyways, can you please try with [this "ponyfill"](https://github.com/sindresorhus/path-is-absolute)? If it works it'll solve our problems just by adding the dependency.

```
npm install path-is-absolute
node
> var pathIsAbsolute = require('path-is-absolute');
> pathIsAbsolute("C:\Users\vagrant\S\fable-samples\consoleSample\solution.fsx");
```

**Comment by davidpodhola** [comment]
Aren't you using forward slashes `/` instead of backslashes `\`? See below, even `path-is-absolute` still returns `false` unless I use forward slashes `/`. Maybe it is necessary as I did in [my commit in Funscript](https://github.com/davidpodhola/FunScript/commit/f25d715ac982882311edb3dac8c91f1d966866ce). 

``````
> var pathIsAbsolute = require('path-is-absolute');
undefined
> pathIsAbsolute("C:\Users\vagrant\S\fable-samples\consoleSample\solution.fsx");

false
> pathIsAbsolute("C:/Users/vagrant/S/fable-samples/consoleSample/solution.fsx");

true
> var q = "C:/Users/david.podhola/S/fable-samples/consoleSample/solution.fsx"
undefined
> var path = require("path")
undefined
> path.isAbsolute(q)
true
> path.resolve(q) == path.normalize(q)
true
>

```

The problem is they [seem to fix it in node.js now](https://github.com/nodejs/node/blob/master/lib/path.js#L250), but did not since at my machine


```
> var q = "C:\Users/david.podhola/S/fable-samples/consoleSample/solution.fsx"
undefined
> code = q.charCodeAt(2)
**85**
>

**Comment by davidpodhola** [comment]
Anyway long story short using `projFile` from outside of the project like 
`node tools\fable2babel.js --projFile "C:\Users\david.podhola\S\fable-samples\consoleSample\solution.fsx"`
produces 
`C:\Users\david.podhola\S\fable-samples\consoleSample>cmd /C C:\Users\david.podhola\S\Fabel\build\main\Fable.exe --lib ..\..\Fabel\lib --outDir . --watch false --projFile solution.fsx --env node` for me (note I am using [the code including my "fix"](https://github.com/fsprojects/Fable/pull/32)) and so I assume it works.

**Comment by alfonsogarciacaro** [comment]
@fdcastel can you please check if the problem is fixed now?

**Comment by fdcastel** [comment]
**TL;DR: Yeah! Its working now!**

Thanks for all the work and sorry for the delay guys (I'm playing with Fable on my free time ;) )

For the sake of completion, see below the results of my tests as asked (running on Windows 8.1, upgraded node to 5.7.0)


```
C:\Temp\Fable>node
> var p = "c:\temp\fable\sample\node\server\app.fsx"
var p = "c:\temp\fable\sample\node\server\app.fsx"
undefined
> var path = require("path")
undefined
> path.isAbsolute(p)
false
> path.resolve(p) == path.normalize(p)
false
```

as you can see, the same results got by @davidpodhola in Windows 10.

HOWEVER, changing to forward slashes gives me:


```
C:\Temp\Fable>node
> var p = "c:/temp/fable/sample/node/server/app.fsx"
var p = "c:/temp/fable/sample/node/server/app.fsx"
undefined
> var path = require("path")
undefined
> path.isAbsolute(p)
true
> path.resolve(p) == path.normalize(p)
true
```

and then


```
C:\Temp\Fable>npm install path-is-absolute
fable@1.0.0 C:\Temp\Fable
`-- path-is-absolute@1.0.0


C:\Temp\Fable>node
> var pathIsAbsolute = require('path-is-absolute');
var pathIsAbsolute = require('path-is-absolute');
undefined
> pathIsAbsolute("c:\temp\fable\sample\node\server\app.fsx");
false
> pathIsAbsolute("c:/temp/fable/sample/node/server/app.fsx");
true
```

Then, I updated my repo with the latest sources and it is fully working now:

```
C:\Temp\Fable>node tools\fable2babel.js --projFile "c:\temp\fable\sample\node\se
rver\app.fsx"
c:\temp\fable\sample\node\server>cmd /C C:\Temp\Fable\build\main\Fable.exe --lib
 ../../../lib --outDir . --watch false --projFile ./app.fsx --env node
Compiled c:\temp\fable\sample\node\server\app.fsx
Finished
```

Nice! ;)

**Comment by alfonsogarciacaro** [comment]
Great, thanks a lot you both for time!
